### PR TITLE
Add cluster scan implementation

### PR DESCRIPTION
**Pull request by estepnv**
Fixes redis/redis-rb#1045

### CODE REVIEW

**Comment by estepnv** [comment]
please suggest what additional tests should I add

File: lib/redis/cluster.rb
Comment (supercaracal) [inline lib/redis/cluster.rb:29]: [nitpick] Rubocop says:

```
Offenses:

lib/redis/cluster.rb:29:34: C: Layout/SpaceAfterComma: Space missing after comma.
      @scan_state = Hash.new { |h,k| h[k] = Hash.new { |h,k| h[k] = {} } }
                                 ^
lib/redis/cluster.rb:29:57: W: Lint/ShadowingOuterLocalVariable: Shadowing outer local variable - h.
      @scan_state = Hash.new { |h,k| h[k] = Hash.new { |h,k| h[k] = {} } }
                                                        ^
lib/redis/cluster.rb:29:58: C: Layout/SpaceAfterComma: Space missing after comma.
      @scan_state = Hash.new { |h,k| h[k] = Hash.new { |h,k| h[k] = {} } }
                                                         ^
lib/redis/cluster.rb:29:59: W: Lint/ShadowingOuterLocalVariable: Shadowing outer local variable - k.
      @scan_state = Hash.new { |h,k| h[k] = Hash.new { |h,k| h[k] = {} } }
                                                          ^
```

Comment (estepnv) [reply lib/redis/cluster.rb]: fixed

File: lib/redis/cluster.rb
Comment (supercaracal) [reply lib/redis/cluster.rb:246]: Does it leak when client quits scanning half-way?

```
irb(main):011:0> r = Redis.new cluster: %w[redis://127.0.0.1:7000]
=> #<Redis client v4.5.1 for redis://127.0.0.1:7000/0 redis://127.0.0.1:7001/0 redis://127.0.0.1:7002/0>

irb(main):012:0> r.keys '*'
=> ["key1", "key2", "{key}1", "{key}2"]

irb(main):013:0> r.scan 0, count: 1
=> ["x-2f620d-61490425", ["key2", "key1", "{key}2"]]

irb(main):014:0> r.instance_variable_get(:@client).instance_variable_get(:@scan_state)
=> {"2f620d"=>{"127.0.0.1:7000"=>{"x-2f620d-61490425"=>"0"}, "127.0.0.1:7001"=>{"x-2f620d-61490425"=>"3"}, "127.0.0.1:7002"=>{"x-2f620d-61490425"=>"1"}}}

irb(main):015:0> r.scan 'x-2f620d-61490425'
=> ["x-2f620d-55435333", ["{key}1"]]

irb(main):016:0> r.instance_variable_get(:@client).instance_variable_get(:@scan_state)
=> {"2f620d"=>{"127.0.0.1:7000"=>{}, "127.0.0.1:7001"=>{"x-2f620d-55435333"=>"0"}, "127.0.0.1:7002"=>{"x-2f620d-55435333"=>"0"}}}

irb(main):017:0> r.scan 'x-2f620d-55435333'
=> ["0", []]

irb(main):018:0> r.instance_variable_get(:@client).instance_variable_get(:@scan_state)
=> {}
```

Comment (supercaracal) [reply lib/redis/cluster.rb:246]: ref: https://redis.io/commands/scan#terminating-iterations-in-the-middle

Comment (estepnv) [reply lib/redis/cluster.rb:246]: or use a dumb solution like this
```ruby
# typed: true
# frozen_string_literal: true

class LruCache
  def initialize(max = 10)
    @cache = {}
    @cache_keys = []
    @size = 0
    @max_size = max
  end

  def fetch(key)
    val = get(key)

    if val.nil?
      if block_given?
        val = yield
        set(key, val)
      end
    end

    sweep

    val
  end

  def sweep
    if @size > @max_size
      @cache.delete(@cache_keys.pop)
      @size -= 1
    end
  end

  def set(key, val)
    @cache[key] = val
    @cache_keys.prepend(@cache_keys.delete(key) || key)
    @size += 1
  end

  def up(key)
    if @cache_keys[0] != key
      @cache_keys.prepend(@cache_keys.delete(key))
    end
  end

  def get(key)
    val = @cache[key]

    up(key) if !val.nil?

    val
  end

  def del(key)
    @cache.delete(key)
    @cache_keys.delete(key)
    @size -= 1
  end
end

```

Comment (estepnv) [reply lib/redis/cluster.rb:246]: or we could prohibit `#scan` usage and raise error. and force users to always use the `#Xscan_each` , to make sure and cleans the state object

**Comment by supercaracal** [comment]
https://github.com/redis/redis/issues/1962#issuecomment-568483242

Redis server maintainer said:

> I don't think that `--scan` and `--cluster` modes play nicely together.

I don't know that how much the feature is needed.
But current redis gem's behavior is a bug indeed. Randomly assign is nonsense.

https://github.com/RedisLabs/redis-cluster-proxy#commands-that-act-differently-from-standard-redis-commands-or-that-have-special-behavior

> SCAN: performs the scan on all the master nodes of the cluster.
> The cursor contained in the reply will have a special four-digits suffix indicating the index of the node that has to be scanned. 
> Note: sometimes the cursor could be something like "00001", so you mustn't convert it to an integer when your client has to use it to perform the next scan.

The above proxy feature supports `SCAN` command by similar interface.
But it seems that the implementation is more simply.

https://github.com/RedisLabs/redis-cluster-proxy/blob/ac83840d1108901a475e24232b7a0986e97e65dd/src/proxy.c#L812-L819

> Scan all master nodes in the cluster.
> Node order is alphabeticallty and it's taken from `clusterGetMasterNames`.
> The index of the node is taken from the cursor last 4 digits.
> If the index is beyond the list of all the master the choosen node will be the last one.
> If the cursor is '0', the choosen node will be the first one.
> This special cursor format will be handled by `handleScanReply`, that will append the index of the node to scan to the cursor replied from the cluster.
> The real cursor sent to the cluster won't actually contain the index suffix.

https://redis.io/commands/scan
Also, It seems that redis server has just a state as a cursor.

Is it better if we implement it by stateless logic?

File: lib/redis/cluster.rb
Comment (supercaracal) [reply lib/redis/cluster.rb:246]: I'd say that the latter solution might be better.
https://github.com/redis/redis-rb/blob/0c43a516bd90127850031ec970e11f343f63fbaf/lib/redis.rb#L3023-L3032

We can simply iterate per nodes without any state.

Comment (estepnv) [reply lib/redis/cluster.rb:246]: by latter you meant stateless, right? as far as i understand, the idea is to try to preserve the state within a returned cursor and we iterate nodes one by one. I'll see what I can do.

**Comment by supercaracal** [comment]
I am just a contributor. It is eventually needed to review by maintainer.

File: lib/redis/cluster.rb
Comment (supercaracal) [reply lib/redis/cluster.rb:246]: > by latter you meant stateless, right?

Yes, it is.

Comment (supercaracal) [reply lib/redis/cluster.rb:246]: I am sorry for my bad English. My "stateless" means cursor only. It seems that maintaining any additional state might be a bit complex.

**Comment by supercaracal** [comment]
### approach 1
* Client raises an error when calling public `#scan` method in cluster mode.
* Only public `#scan_each` method is legitimate when using cluster mode.
* It needs conditional branching in public `#scan_each` method.

### approach 2
* Client knows sorted node keys such that `host:port`.
  * ex) `127.0.0.1:7000, 127.0.0.1:7001, ...`
* At first, we can call public `#scan` method with `0`.
* Client returns next cursor and scanned keys. The cursor includes current node key index.
  * ex) `17-0001, 0-0002, 42-0002, 0-0003, ...`
* Client can iterate every nodes using only cursor value.

File: lib/redis/cluster.rb
Comment (estepnv) [inline lib/redis/cluster.rb:243]: @supercaracal pls take a look. 
so it's pretty simple
I use last 8 bits to encode the client index the rest is the raw key
I iterate keys on a node until I receive `"0"` cursor, after that I increment the index
if current client is nil it's the return condition

Comment (estepnv) [reply lib/redis/cluster.rb:242]: it's the second approach you proposed but without strings

File: lib/redis/cluster.rb
Comment (supercaracal) [inline lib/redis/cluster.rb:248]: It might be readable if this step is added as a public method to `Redis::Cluster::Node` class.

```ruby
class Redis
  class Cluster
    class Node
      def scale_reading_clients
        @clients.select { |k, v| replica_disabled? ? master?(k) : slave?(k) }.values
      end
    end
  end
end
```

File: lib/redis/cluster.rb
Comment (supercaracal) [reply lib/redis/cluster.rb:242]: Thank you for your improvement.

**Review by supercaracal — APPROVED** [review]
:+1:

File: lib/redis/cluster.rb
Comment (byroot) [inline lib/redis/cluster.rb:243]: Maybe use `Integer` here? So that it raises if we get something that's not an integer.

File: lib/redis/cluster.rb
Comment (byroot) [inline lib/redis/cluster.rb:256]: Same here. `Integer` would make sense.

File: lib/redis/cluster.rb
Comment (byroot) [inline lib/redis/cluster.rb:245]: Probably not much, but we could directly use `256` to not repeatedly recompute that exponent.

**Review by byroot — COMMENTED** [review]
Please squash your commits together and I'll merge.

File: lib/redis/cluster.rb
Comment (estepnv) [reply lib/redis/cluster.rb]: @byroot updated

File: lib/redis/cluster.rb
Comment (estepnv) [inline lib/redis/cluster.rb:140]: @byroot what about zscan/hscan/sscan, will it work? may I should add more tests? I'm not very familiar with internals.

Comment (byroot) [reply lib/redis/cluster.rb:140]: More tests wouldn't hurt.

**Pull request merged**
